torchtune/datasets/_packed.py

```diff
@@ -0,0 +1,113 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import Dict, List, Optional, Tuple
+
+from torch.utils.data import Dataset
+from tqdm import tqdm
+
+
+class PackedDataset(Dataset):
+    """
+    Performs greedy sample packing on a provided dataset. This is done as a single
+    preprocessing step before training begins. Shuffling is done outside of this
+    class on packed samples as part of the dataloader.
+
+    The class loads, tokenizes, and packs examples on initialization - no tokenization is done during training.
+
+    The general flow on initialization is: load tokenized sample -> add to buffer ->
+        when buffer is long enough, add to self.samples.
+
+    During training, returns self.samples[idx] as input and label.
+
+    Args:
+        ds (Dataset): dataset to sample pack. This should return a tuple of tokenized
+            inputs and labels.
+        max_seq_len (int): Maximum number of tokens to pack
+        max_rows (Optional[int]): maximum number of samples to pack. Default is None, which will pack as many samples as possible.
+        split_samples (bool): if the last sample in a pack does not fit in ``max_seq_len``,
```

Comment:
I feel like ```split_across_instances``` or ```split_across_boundary``` would be more intuitive. If I just found this flag in a config, I'd interpret it as splitting all samples across instances i.e. no packing. Ignore this comment if this is an accepted convention in the community. If not, I'd suggest thinking a bit about this flag.

Reply:
I'll change it to `split_across_pack`